README.txt
RapidMiner
 
 Copyright (C) 2001-2006 by Rapid-I and contributors
 
 Complete list of developers available at our web site:

 http://rapid-i.com

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as 
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version. 

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
USA.

--------------------------------------------------------------------

Please note:
------------

* The license of all libraries can be found in the directory 
  "licences" located in the main directory of RapidMiner.
  
* The only other libraries under GPL are Weka and JUnit.

* The library "felt.jar" is not licensed under the GPL and
  it is NOT allowed to use it for other applications.

* It is NOT allowed to use the icons of RapidMiner for other
  applications.
  
* All other libraries are licensed under LGPL or BSD-style
  licenses.
  
* RapidMiner itself (without the library "felt.jar", see above) is 
  licensed under the GPL (see the corresponding license file).
  
 
Library Notes:
--------------

Along with the core RapidMiner classes you have received some software
packages and libraries not all of which are subject to the RapidMiner
license terms. You can find the license texts of these libraries
in the 'licenses' directory of RapidMiner. These libraries are

  src/libsvm/*	The java source files are part of the libsvm software
		by Chih-Chung Chang and Chih-Jen Lin, who hold the
		Copyright 2000-2005. The software can be downloaded
		from

                   http://www.csie.ntu.edu.tw/~cjlin/libsvm/

        We have made some minor modifications (removed
		System.out messages, added a method for writing the
		model to a stream, renamed some classes). You can find the 
		libsvm license in README.libsvm.
		
		Current Version: own based on 2.84

  lib/weka.jar  This java file is part of the Weka machine learning 
                software which has its home at

		   http://www.cs.waikato.ac.nz/ml/weka/

		Weka comes under the GNU General Public License and
		the original copyright note is

		Copyright (C) 1998, 1999, 2000, 2001  Eibe Frank, 
		   Leonard Trigg, Mark Hall, Richard Kirkby

		You can find the Weka original readme file in
		README.weka
		
		Current Version: 3.5.6 (latest CVS build)

  lib/colt.jar  Colt is a library for scalable scientific and technical 
                computing in Java. It has its home at
                
                   http://dsd.lbl.gov/~hoschek/colt/
                   
                Current Version: 1.2.0

  lib/Jama.jar  JAMA is a basic linear algebra package for Java. This 
		archive holds the classes of the JAva MAtrix package 
		which is a cooperative product of The MathWorks and
		the National Institute of Standards and Technology 
		(NIST) which has been released to the public domain.
		The software can be downloaded from

		   http://math.nist.gov/javanumerics/jama/
		   
		Current Version: 1.0.2

  lib/jung-algorithms.jar
  lib/jung-api.jar
  lib/jung-graph-impl.jar
  lib/jung-io.jar
  lib/jung-visualization.jar
                JUNG is the Java Universal Network/Graph Framework --
                a software library that provides a common and extendible 
                language for modeling, analysis, and visualization of 
                data. It has its home at

		           http://jung.sourceforge.net/
		           
		        Current Version: 2.0alpha2

  lib/commons-collections.jar
          
                The Jakarta commons library provides some reusable 
                Java components. It has its home at

                   http://jakarta.apache.org/commons/
                   
                It is needed by xstream.
                   
                Current Version: 3.2
                
  lib/collections-generic.jar
          
                The Jakarta generic collections library provides 
                some reusable Java generic collections components. 
                It has its home at

                   http://jakarta.apache.org/commons/
                   
                It is needed by Jung.
                   
                Current Version: 4.0.1
                
   lib/concurrent.jar
          
                This package provides standardized, efficient versions
                of utility classes commonly encountered in concurrent
                Java programming. It has its home

                   http://g.oswego.edu/dl/classes/EDU/oswego/cs/dl/util/concurrent/intro.html
                   
                It is needed by Jung.
                   
                Current Version: 1.3.4

  lib/junit.jar  JUnit is a testing framework. Please note that JUnit is only
                 used during development and is not part of the non-GPL releases
                 Is has its home at

                    http://www.junit.org/index.htm
                    
                 Current Version: 4.3.1
                   
  lib/jmathplot.jar  JMathPlot is a plotter framework which is used for the
                     3D plots and some of the 2D plots. Is has its home at

                       http://jmathtools.sourceforge.net/
                       
                     Current Version: 2006-09-15          

  lib/xstream.jar  XStreams is a very convenient library for creating XML
                   based serializations for arbitrary Java objects. It can
                   be found at

                    http://xstream.codehaus.org
                   
                   Current Version: 1.2.2

  lib/xpp3.jar  XPP3 is a fast XML pull parser used by XStreams. It is 
                part of the XStreams package.
                
                Current Version: 3
  
  lib/felt.jar  This library contains the look and feel of RapidMiner. 
                Very important: This library is not part of the GPL
                open-source version of RapidMiner and is not allowed to
                be used for other applications. The same applies for
                the icons of RapidMiner.
                
                Current Version: own based on 1.8.2
                
  lib/jugpreview.jar  Contains the Print Preview dialog
  
                      Current Version: own
                
  lib/jxl.jar  JExcel is used to load Microsoft Excel spreadsheets into
               RapidMiner as data input.
               
               Current Version: 2.6.2
               
  lib/joone.jar  A library for neural network learning. This library is
                 changed from the original by introducing the usage of 
                 fixed random numbers and some other small modifications.
  
                 Current Version: own based on 1.2.1

  lib/jfreechart.jar  A library for painting some of the charts available
                      in RapidMiner
  
                 Current Version: 1.0.5


  lib/jcommon.jar  This library is used by the jfreechart library
  
                 Current Version: 1.0.9
                                 
  lib/freehep/*	 These jar files are part of the Java FreeHEP
                 library. It is used for PDF and PS file export.
		         The Copyright 2000 holders are:
		
		            CERN, Geneva, Switzerland
		            SLAC, Stanford, California, U.S.A.
                    University of California Santa Cruz, U.S.A.

		         This software is distributed under the LGPL license, see
		         the file lib/freehep/LGPL.txt. It can be downloaded from

		            http://www.freehep.org/
		            
		         Current Version: 2.0

CHANGES.txt
History of changes of the RapidMiner / Yale versions
====================================================

Changes from RapidMiner 4.0beta2 to RapidMiner 4.0beta3 [????/??/??]
-------------------------------------------------------------------

  * New operators:
      - SimplePerformance (could be used in most cases instead of 
        now deprecated PerformanceEvaluator)
      - ClassificationPerformance
      - BinominalClassificationPerformance
      - RegressionPerformance
      - UserBasedPerformance
      - SingleRuleWeighting
        
  * Almost all process setups will now also correctly work if the 
    nominal values of training and test data are not defined or 
    are not defined in the same order
  
  * The somewhat big operator "PerformanceEvaluator" is now
    deprecated and was divided into several smaller operators
    which now fit the different learning task types. 
      
  * Added a filter for the New Operator tab
  
  * Added learning for numerical attributes for rule learners
  
  * Renamed lowest verbosity level to "all"
  
  * Improved visualization of performance criteria
  
  * Improved ExampleSetMerge
  
  * Improved rule learning on numerical data sets
  
  * Improved tree learning on numerical data sets
  
  * Added single example visualization after clicking a node
    in the graph view of similarity visualizations
    
  * Implementation Details:
      - Revised clustering operators and introduced improved
        abstract clustering
      - The global logging can now be specified either by
        general properties or via the method
        LogService.initGlobalLogging(...)
      - Attribute.getStatistics(...) is now deprecated, please
        use ExampleSet.getStatistics(Attribute, ...) instead
      - Changed the log verbosity of the process informations
        at the beginning and the end of process executions
     
  * Bugfixes:
      - Removed unnecessary parameters from RandomForest
      - Fixed attribute name bug (not case sensitive) causing
        errors in some preprocessing operators if features with 
        the same name but different cases exists
      - Fixed bug in Anova and T-Test calculation (wrong degree
        of freedom)
      - Fixed bug during weight normalization which lead in many
        cases to a concurrent modification exception which was
        covered by a process change message
      - Removed possible bug in UPGMA-Clustering
      - Graph View of Cluster Models did not work
      - Added missing clone in discretization operator which 
        might have caused problems in cases where the discretization
        was added into an iterating chain (like validation chains)
      - Streams are now not automatically closed during XML (de-)
        serialization
      - Rule learners did not produce greater equal conditions
        
        
Changes from RapidMiner 4.0beta to RapidMiner 4.0beta2 [2007/06/24]
-------------------------------------------------------------------
  
  * New operators:
      - BatchXValidation
      - BatchSlidingWindowValidation
      - AttributeCopy
      - ExampleSetTranspose
      - AssociationRuleGenerator
      - RelevanceTree
      - CHAID
      - Tree2RuleConverter

  * Removed operators:
      - RegressionTree (may be re-added in later releases)
      - Ripper (replaced by RuleLearner)    
  
  * Renamed operators (old operator names are deprecated now):
      - ExperimentEmbedder operator was renamed
        to ProcessEmbedder (see below)
      - ExperimentLog operator was renamed to 
        ProcessLog operator (see below)
          
  * API change: Renamed Experiment to Process 
    (the old class Experiment is still available for
    compatibility reasons but deprecated)
    
  * API change: OperatorService.createOperator(Class)
    is now the preferred way for operator creation
    and does no longer need a cast (generics)
  
  * Added correct file encodings to all IO operators
  
  * Renamed log verbosity "minimum" to "all" and log
    verbosity "maximum" to "off"
    
  * Added meaningful default and range values for the 
    parameters of the ParameterOptimization operators
  
  * Replaced Tip of the Day dialog by the tip in the 
    Welcome screen
  
  * Changed all Weka parameters to non-expert parameters
    (available in beginners mode)
  
  * SVMWeighting now supports more than 2 classes
  

src/StyleCheck.java
/*
 *  RapidMiner
 *
 *  Copyright (C) 2001-2006 by Rapid-I and the contributors
 *
 *  Complete list of developers available at our web site:
 *
 *       http://rapid-i.com
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as 
 *  published by the Free Software Foundation; either version 2 of the
 *  License, or (at your option) any later version. 
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 *  USA.
 */

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.Writer;
import java.text.NumberFormat;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.SortedMap;
import java.util.TreeMap;

import com.rapidminer.tools.Tools;


/**
 * This class can be used to perform some style checks. In addition, this
 * program might give useful informations like the totalClasses number of lines or the
 * classes written by a particular author.
 * 
 * @author Ingo Mierswa
 * @version $Id: StyleCheck.java,v 2.14 2007/05/28 21:23:34 ingomierswa Exp $
 */
public class StyleCheck {

	/**
	 * This command indicates that all classes should be checked for maximal
	 * length.
	 */
	private static final String SIZE_CHECK = "size_check";

	/**
	 * This command indicates that the totalClasses number of all lines of code of all
	 * classes should be counted.
	 */
	private static final String LINES_COUNT = "total_lines";

	/**
	 * This command indicates that all classes should be checked for missing
	 * class comments.
	 */
	private static final String COMMENT_CHECK = "comment_check";

	/**
	 * This command indicates that all classes should be checked for missing
	 * version tags.
	 */
	private static final String VERSION_CHECK = "version_check";

	/**
	 * This command indicates that all classes should be checked for missing
	 * author tags.
	 */
	private static final String AUTHOR_CHECK = "author_check";

	/**
	 * This command indicates that the a list of all classes should be returned
	 * written by the given author.
	 */
	private static final String AUTHOR_LIST = "author_list";

	/**
	 * This command indicates that the number of all classes should be counted
	 * written by the given author.
	 */
	private static final String AUTHOR_COUNT = "author_count";

	/**
	 * This command indicates that the given author name should be replaced by
	 * the new one (e.g. 'ingo' by 'Ingo Mierswa').
	 */
	private static final String AUTHOR_REPLACE = "author_replace";

	/**
	 * This command indicates that a list of all author names should be printed
	 * with the number of class they have written.
	 */
	private static final String ALL_AUTHORS = "all_authors";

	/**
	 * This command indicates that all classes should be checked for the license
	 * text.
	 */
	private static final String LICENSE_CHECK = "license_check";

	/** Superclass for all Java file checker. */
	private static abstract class JavaFileChecker {

		/**
		 * Default implementation does nothing. Invoked after check and might be
		 * overridden by subclasses.
		 */
		public void printResult() {}

		/** Recursively checks all files. */
		public void checkAllFiles(File current) throws Exception {
			if (!current.isDirectory()) {
				if (current.getName().endsWith("java"))
					performCheck(current);
			} else {
				File[] children = current.listFiles();
				for (int i = 0; i < children.length; i++)
					checkAllFiles(children[i]);
			}
		}

		/** Must be implemented by subclasses. */
		public abstract void performCheck(File file) throws Exception;
	}

	/**
	 * Checks all Java files for a maximum size and prints a warning message if
	 * this number of lines is exceeded.
	 */
	private static class SizeChecker extends JavaFileChecker {

		private int maxLines;

		public SizeChecker(int maxLines) {
			this.maxLines = maxLines;
		}

		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			int counter = 0;
			String line = null;
			while ((line = in.readLine()) != null) {
				if (line.trim().length() > 0)
					counter++;
			}
			in.close();
			if (counter > maxLines) {
				System.out.println(file.getPath() + ": more than " + maxLines + " lines!");
			}
		}
	}

	/** Counts all lines and prints the totalClasses number of lines. */
	private static class LineCounter extends JavaFileChecker {

		private int total = 0;
		
		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			while ((line = in.readLine()) != null) {
				if (line.trim().length() > 0)
					total++;
			}
			in.close();
		}

		public void printResult() {
			System.out.println("Total number of lines: " + total);
		}
	}

	/** Checks all Java files for a class comment. */
	private static class CommentChecker extends JavaFileChecker {

		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			boolean commentFound = false;
			boolean inComment = false;
			boolean packageFound = false;
			while ((line = in.readLine()) != null) {
				// ignore everything before the package starts
				if (line.trim().startsWith("package"))
					packageFound = true;
				if (!packageFound)
					continue;

				if (line.trim().startsWith("/**")) {
					commentFound = true;
					inComment = true;
				}
				if (inComment && (line.indexOf("*/") > 0))
					inComment = false;

				if ((!inComment) && ((line.indexOf(" class ") > 0) || (line.indexOf(" interface ") > 0)))
					break;
			}
			in.close();
			if (!commentFound) {
				System.out.println(file.getPath() + ": it seems that there is no class comment (or no package...)!");
			}
		}
	}

	/** Checks all Java files for a version tag. */
	private static class VersionChecker extends JavaFileChecker {

		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			boolean versionFound = false;
			while ((line = in.readLine()) != null) {
				if (line.indexOf("@version") > 0) {
					if (line.indexOf("$Id") > 0) {
						versionFound = true;
						break;
					}
				}
				// if ((line.indexOf(" class ") > 0) || (line.indexOf("
				// interface ") > 0))
				// break;
			}
			in.close();
			if (!versionFound) {
				System.out.println(file.getPath() + ": missing or non-conform @version tag!");
			}
		}
	}

	/** Checks all Java files for an author tag. */
	private static class AuthorChecker extends JavaFileChecker {

		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			boolean authorFound = false;
			while ((line = in.readLine()) != null) {
				if (line.indexOf("@author") > 0) {
					authorFound = true;
					break;
				}
				// if ((line.indexOf(" class ") > 0) || (line.indexOf("
				// interface ") > 0))
				// break;
			}
			in.close();
			if (!authorFound) {
				System.out.println(file.getPath() + ": missing @author tag!");
			}
		}
	}

	/** Collects all Java files written by the given author (and not by the second, optionally). */
	private static class AuthorList extends JavaFileChecker {

		private String author;

        private String exceptAuthor;
        
		private List<String> result = new LinkedList<String>();

		public AuthorList(String author) {
			this(author, null);
		}

        public AuthorList(String author, String exceptAuthor) {
            this.author = author;
            this.exceptAuthor = exceptAuthor;
        }
        
		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			while ((line = in.readLine()) != null) {
                int index = line.indexOf("@author"); 
				if (index >= 0) {
                    String authorLine = line.substring(index + "@author".length());
                    String[] authors = authorLine.split(",");
                    boolean authorFound = false;
                    boolean exceptAuthorFound = false;
                    for (String a : authors) {
                        if (a.trim().toLowerCase().indexOf(author.toLowerCase()) >= 0) {
                            authorFound = true;
                        }
                        if ((exceptAuthor != null) && (a.trim().toLowerCase().indexOf(exceptAuthor.trim().toLowerCase()) >= 0)) {
                            exceptAuthorFound = true;
                        }
                    }
                    if (authorFound && (!exceptAuthorFound))
                        result.add(file.getPath());
                    break;
                }
			}
			in.close();
		}

		public void printResult() {
			System.out.println(result.size() + " classes written by " + author + (exceptAuthor != null ? " and not by " + exceptAuthor + ":" : ":"));
			Iterator<String> i = result.iterator();
			while (i.hasNext()) {
				System.out.println(i.next());
			}
		}
	}

	/** Counts all Java files written by the given author. */
	private static class AuthorCounter extends JavaFileChecker {

		private String author;

		private int counter = 0;

		private int total = 0;

		public AuthorCounter(String author) {
			this.author = author;
		}

		public void performCheck(File file) throws Exception {
			total++;
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			while ((line = in.readLine()) != null) {
				if (line.indexOf("@author") > 0) {
					if (line.toLowerCase().indexOf(author.toLowerCase()) > 0) {
						counter++;
						break;
					}
				}
			}
			in.close();
		}

		public void printResult() {
			System.out.println("Number of classes written by " + author + ": " + counter + " / " + total);
		}
	}

	/** Counts all Java files written by the given author. */
	private static class AuthorReplace extends JavaFileChecker {

		private String oldName, newName;

		private int counter = 0;

		public AuthorReplace(String oldName, String newName) {
			this.oldName = oldName;
			this.newName = newName;
		}

		public void performCheck(File file) throws Exception {
			boolean replace = false;
			StringBuffer result = new StringBuffer();
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			while ((line = in.readLine()) != null) {
				if (line.indexOf("@author") > 0) {
					String newAuthorLine = line;
					if (line.toLowerCase().indexOf(newName.toLowerCase()) < 0) {
						int start = line.toLowerCase().indexOf(oldName.toLowerCase());
						if (start > 0) {
							newAuthorLine = line.substring(0, start) + newName;
							newAuthorLine += line.substring(start + oldName.length());
							replace = true;
						}
					}
					result.append(newAuthorLine + Tools.getLineSeparator());
				} else {
					result.append(line + Tools.getLineSeparator());
				}
			}
			in.close();

			if (replace) {
				counter++;
				Writer out = new FileWriter(file);
				out.write(result.toString().toCharArray());
				out.close();
			}
		}

		public void printResult() {
			System.out.println("Number of classes where " + oldName + " was replaced by " + newName + ": " + counter);
		}
	}

	/** Counts all Java files written by all authors and prints them. */
	private static class AllAuthors extends JavaFileChecker {

        private int totalLines = 0;
        
		private int totalClasses = 0;

		private SortedMap<String, Integer> authorClassNumberMap = new TreeMap<String, Integer>();

		private SortedMap<String, Integer> authorLineNumberMap = new TreeMap<String, Integer>();

		public void performCheck(File file) throws Exception {
			totalClasses++;
			String[] authors = null;
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			int lines = 0;
			while ((line = in.readLine()) != null) {
				lines++;
				int index = line.indexOf("@author");
				if (index > 0) {
					String authorLine = line.substring(index + "@author".length());
					authors = authorLine.split(",");
				}
			}
            totalLines += lines;
			if (authors != null) {
				for (int i = 0; i < authors.length; i++) {
					String currentAuthor = authors[i].trim();

					Integer classCounter = authorClassNumberMap.get(currentAuthor);
					if (classCounter == null) {
						authorClassNumberMap.put(currentAuthor, 1);
					} else {
						authorClassNumberMap.put(currentAuthor, classCounter.intValue() + 1);
					}

					Integer lineCounter = authorLineNumberMap.get(currentAuthor);
					if (lineCounter == null) {
						authorLineNumberMap.put(currentAuthor, lines);
					} else {
						authorLineNumberMap.put(currentAuthor, lineCounter.intValue() + lines);
					}
				}
			}
			in.close();
		}

		public void printResult() {
			System.out.println("Total number of classes: " + totalClasses);
            System.out.println("Total number of lines:   " + totalLines);
            System.out.println("----------------------------------------------------------");
            
            class AuthorCounter implements Comparable<AuthorCounter> {
                private String name;
                private int lines;
                private int classes;
                private NumberFormat formatter;
                
                public AuthorCounter(String name, int lines, int classes, NumberFormat formatter) {
                    this.name = name;
                    this.lines = lines;
                    this.classes = classes;
                    this.formatter = formatter;
                }
                public int compareTo(AuthorCounter a) {
                    return -1 * Double.compare(this.classes, a.classes);
                }
                public boolean equals(Object o) {
                	if (!(o instanceof AuthorCounter)) {
                		return false;
                	} else {
                		return this.name.equals(((AuthorCounter)o).name);
                	}
                }
                public int hashCode() {
                	return this.name.hashCode();
                }
                public int getStringLength() {
                    return (name + ": " + classes + " (" + lines + " lines)").length();
                }
                public String toString(int numberOfSpaces) {
                    StringBuffer spaceString = new StringBuffer();
                    for (int i = 0; i < numberOfSpaces; i++)
                        spaceString.append(" ");
                    double classPercent = (double)classes / (double)totalClasses; 
                    double linesPercent = (double)lines / (double)totalLines; 
                    return 
                        name + ": " + classes + " (" + lines + " lines) " + spaceString + " --> " + 
                        formatter.format(classPercent) + " (" + formatter.format(linesPercent) + ")";
                }
            }
            
            NumberFormat formatter = NumberFormat.getPercentInstance();
            formatter.setMaximumFractionDigits(1);
            formatter.setMinimumFractionDigits(1);
            List<AuthorCounter> authorCounters = new LinkedList<AuthorCounter>();
            Iterator<String> i = authorClassNumberMap.keySet().iterator();
            int maxLength = 0;
			while (i.hasNext()) {
				String author = i.next();
                int classNumber =  authorClassNumberMap.get(author);
                int lineNumber  = authorLineNumberMap.get(author);
                AuthorCounter newCounter = new AuthorCounter(author, lineNumber, classNumber, formatter); 
                authorCounters.add(newCounter);
                maxLength = Math.max(maxLength, newCounter.getStringLength());
			}
            Collections.sort(authorCounters);
            Iterator<AuthorCounter> a = authorCounters.iterator();
            while (a.hasNext()) {
                AuthorCounter current = a.next();
                System.out.println(current.toString(maxLength - current.getStringLength()));
            }
            
            System.out.println("----------------------------------------------------------");
		}
	}

	/** Checks all Java files for a version tag. */
	private static class LicenseChecker extends JavaFileChecker {

		private static String LICENSE_LINE = "Copyright (C) 2001-2006";

		public void performCheck(File file) throws Exception {
			BufferedReader in = new BufferedReader(new FileReader(file));
			String line = null;
			boolean licenseFound = false;
			while ((line = in.readLine()) != null) {
				if (line.indexOf(LICENSE_LINE) > 0) {
					licenseFound = true;
					break;
				}
			}
			in.close();
			if (!licenseFound) {
				System.out.println(file.getPath() + ": missing or non-conform license");
			}
		}
	}

	// ================================================================================

	public static void checkSize(String[] argv) throws Exception {
		if (argv.length != 3) {
			System.err.println("usage: java StyleCheck " + SIZE_CHECK + " max_lines path");
			System.exit(1);
		}
		JavaFileChecker checker = new SizeChecker(Integer.parseInt(argv[1]));
		checker.checkAllFiles(new File(argv[2]));
		checker.printResult();
	}

	public static void countLines(String[] argv) throws Exception {
		if (argv.length != 2) {
			System.err.println("usage: java StyleCheck " + LINES_COUNT + " path");
			System.exit(1);
		}
		JavaFileChecker checker = new LineCounter();
		checker.checkAllFiles(new File(argv[1]));
		checker.printResult();
	}

	public static void checkComment(String[] argv) throws Exception {
		if (argv.length != 2) {
			System.err.println("usage: java StyleCheck " + COMMENT_CHECK + " path");
			System.exit(1);
		}
		JavaFileChecker checker = new CommentChecker();
		checker.checkAllFiles(new File(argv[1]));
		checker.printResult();
	}

	public static void checkVersion(String[] argv) throws Exception {
		if (argv.length != 2) {
			System.err.println("usage: java StyleCheck " + VERSION_CHECK + " path");
			System.exit(1);
		}
		JavaFileChecker checker = new VersionChecker();
		checker.checkAllFiles(new File(argv[1]));
		checker.printResult();
	}

	public static void checkAuthor(String[] argv) throws Exception {
		if (argv.length != 2) {
			System.err.println("usage: java StyleCheck " + AUTHOR_CHECK + " path");
			System.exit(1);
		}
		JavaFileChecker checker = new AuthorChecker();
		checker.checkAllFiles(new File(argv[1]));
		checker.printResult();
	}

	public static void createAuthorList(String[] argv) throws Exception {
		if ((argv.length < 3) || (argv.length > 4)) {
			System.err.println("usage: java StyleCheck " + AUTHOR_LIST + " path authorname [exceptauthor]");
			System.exit(1);
		}
		JavaFileChecker checker = null;
        if (argv.length == 3)
            checker = new AuthorList(argv[2]);
        else
            checker = new AuthorList(argv[2], argv[3]);
        
		checker.checkAllFiles(new File(argv[1]));
		checker.printResult();
	}

	public static void createAuthorCount(String[] argv) throws Exception {
		if (argv.length != 3) {
			System.err.println("usage: java StyleCheck " + AUTHOR_COUNT + " authorName path");
			System.exit(1);
		}
		JavaFileChecker checker = new AuthorCounter(argv[1]);
		checker.checkAllFiles(new File(argv[2]));
		checker.printResult();
	}

	public static void replaceAuthor(String[] argv) throws Exception {
		if (argv.length != 4) {
			System.err.println("usage: java StyleCheck " + AUTHOR_REPLACE + " oldAuthorName newAuthorName path");
			System.exit(1);
		}
		JavaFileChecker checker = new AuthorReplace(argv[1], argv[2]);
		checker.checkAllFiles(new File(argv[3]));
		checker.printResult();
	}

	public static void listAllAuthors(String[] argv) throws Exception {
		if (argv.length != 2) {
			System.err.println("usage: java StyleCheck " + ALL_AUTHORS + " path");
			System.exit(1);
		}
		JavaFileChecker checker = new AllAuthors();
		checker.checkAllFiles(new File(argv[1]));
		checker.printResult();
	}

	public static void checkForLicenseText(String[] argv) throws Exception {
		if (argv.length != 2) {
			System.err.println("usage: java StyleCheck " + LICENSE_CHECK + " path");
			System.exit(1);
		}
		JavaFileChecker checker = new LicenseChecker();
		checker.checkAllFiles(new File(argv[1]));
	}

	// ================================================================================

	public static void main(String[] argv) throws Exception {
		if (argv.length < 2) {
			System.err.println("usage: java StyleCheck command <arguments> path");
			System.exit(1);
		}
		if (argv[0].equals(SIZE_CHECK)) {
			checkSize(argv);
		} else if (argv[0].equals(LINES_COUNT)) {
			countLines(argv);
		} else if (argv[0].equals(COMMENT_CHECK)) {
			checkComment(argv);
		} else if (argv[0].equals(VERSION_CHECK)) {
			checkVersion(argv);
		} else if (argv[0].equals(AUTHOR_CHECK)) {
			checkAuthor(argv);
		} else if (argv[0].equals(AUTHOR_LIST)) {
			createAuthorList(argv);
		} else if (argv[0].equals(AUTHOR_COUNT)) {
			createAuthorCount(argv);
		} else if (argv[0].equals(AUTHOR_REPLACE)) {
			replaceAuthor(argv);
		} else if (argv[0].equals(ALL_AUTHORS)) {
			listAllAuthors(argv);
		} else if (argv[0].equals(LICENSE_CHECK)) {
			checkForLicenseText(argv);
		} else {
			System.err.println("Unknown command: " + argv[0]);
			System.exit(1);
		}
	}
}
